compiler/ast/src/types/type_.rs

```diff
@@ -94,6 +96,14 @@ impl Type {
             (Type::Composite(left), Type::Composite(right)) => {
                 left.id.name == right.id.name && left.program == right.program
             }
+            (Type::Future(left), Type::Future(right)) if left.inputs.len() == right.inputs.len() => {
```

Comment:
As a safety check, should we require that the location is populated?

Reply:
If both are `None` should that fail then? Since the location is not populated by the parser, many future type instances do not have a location.

Reply:
Would they be inferred before they are checked for equality?